articles/141_static_remapping.md

```diff
@@ -0,0 +1,341 @@
+Client libraries also have APIs in code to pass remapping rules when the node is initialized.
+A remap rule consists of two names: one that should be replaced with another.
+
+ROS 1 remapping works on **Fully Qualified Names** (FQN).
+Both sides of a rule are [expanded to FQN](http://docs.ros.org/api/roscpp/html/namespaceros_1_1names.html#a377ff8fede7b95398fd5d1c5cd49246b).
+Before a name is remapped it is also [expanded to FQN](http://docs.ros.org/api/roscpp/html/namespaceros_1_1names.html#ab2eebaf734abfbdccb4122f8e24f547f).
+The name is remapped to the right side only if it exactly matches the left side of a rule.
+
+
+## Remapping rule syntax
+This is a proposal for the ROS 2 remapping rule syntax.
+It attempts to be the same as ROS 1 syntax when possible, and different where it supports new features.
+
+Requirements supported by this syntax
+- Remap One Node in a Process
+- Change a Namespace
+- Change a Basename
+- Exact FQN Replacement
+- Exact Relative Name Replacement
+- Remap via Command Line
+- Change the Default Namespace
+
+Requirements not supported
+- Change a Token
+- Pre-FQN Remapping
+
+Special Operators
+- `*` wild card that matches a single token
+- `**` wild card that matches at least one slash and zero or more tokens
+- `:=` divides the two parts of a remapping rule: matching and replacement
+- `~` expands to `/namespace/nodename` and may only be at the very beginning of a match or replacement
+- `__ns` matches the default namespace if used by itself on the match side of a rule
+- `nodename:` prefixed to a rule makes it apply only to a node with that name
+- `\1` - `\9` are replaced with the matched content of a wild card
+
+The operators `*` and `**` are chosen to match globbing behavior in bash.
+`**` behaves similarly as it does in bash>=4.0 with the globstar option set.
+The syntax for `\[1-9]` was taken from the syntax for backreferences in POSIX BRE.
+The operators `~`, `:=`, and `__ns` have the same meaning in ROS 1.
+
+### Applications of the syntax
+The following sections explain how the syntax enables the use cases above.
+
+#### Supporting: Remap One Node in a Process
+Remapping a node in a process requires a way to uniquely identify a node.
+Assuming the node's name is unique in a process, a rule can be prefix with the name and a `:`.
+If the node name is not prefixed, the rule should be applied to all nodes in the process.
+
+*Example*
+
+- Multiple nodes in a process use the name `scan`
+- A rule is given to one node `node1:scan:=scan_filtered`
+- Only node1 uses the rule
+
+
+#### Supporting: Change a Namespace
+There are two cases: changing part of a namespace, and changing the entire namespace.
+The first case requires a wildcard to match the rest of a namespace.
+The second requires a wildcard to match the basename at the end.
+
+*Example: Partial namespace replacement*
+
+- Node uses names `/foo`, `/foo/bar/baz`
+- Node given rule `/foo**:=/fizz\1`
```

Comment:
@dirk-thomas I see what you mean about it feeling more natural. Without the slash it looks like it would match `/foobar`, but I didn't intend that. 

@wjwwood If `**` matched zero or more slashes and whole tokens only, the distinction of `\1` being `/bar` vs `bar` could be made unimportant by the replacement side of a rule automatically adding or removing slashes. For example if `\1` is `/bar` and the replacement rule is `/\1` the result would be `/bar` (slash auto-deleted), and if `\1` is `bar` and the replacement rule is `/foo\1` the result would be `/foo/bar` (slash auto-added).